### PR TITLE
Better http error handling

### PR DESCRIPTION
**Pull request by Phocea**
Modified the way ng-admin is handling HTTP errors so it can easily be customised:

- Renamed ErrorHandler to HttpErrorHandler to avoid confusion.
- Created a HttpErrorService in charge of taking the appropriate action depending on the HTTP code error.
- Added default behavior for 403 error (as discussed with @jpetitcolas )
- Added example of using Angular decorator to customise the service.

### CODE REVIEW

**Comment by Phocea** [comment]
Fix  #1031 
Fix #917

File: doc/Theming.md [hunk 1]
```diff
@@ -181,3 +181,21 @@ entity.errorMessage(function (response) {
     return 'Global error: ' + response.status + '(' + response.data + ')';
 });
 ```
+## Customizing HTTP Error Messages
+If you want to override, patch or extend the way HTTP error are handled by the ng-admin Http Error Service (../src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js) then you may use an [Angular decorator](https://docs.angularjs.org/guide/decorators), as per the below example:
```
Comment (jpetitcolas) [inline doc/Theming.md]: s/HTTP error/HTTP errors/

Comment (jpetitcolas) [inline doc/Theming.md]: Can you fix the link on `Http Error Service`? 

@Kmaschta: seems sedy failed. Moreover, would it work if I try to fix the link (with slashes in it)?

File: doc/Theming.md [hunk 2]
```diff
@@ -181,3 +181,21 @@ entity.errorMessage(function (response) {
...
+```js
+// Change HTTP 403 error notification to display them as information and not errors ('humane-flatty-info' instead of 'humane-flatty-error')
+myApp.config([ '$provide', function($provide) {
+  $provide.decorator('httpErrorService', ['$delegate', '$translate', 'notification', function $httpErrorDecorator($delegate, $translate, notification) {
+      var handle_403_error = $delegate.handle_403_error;
```
Comment (jpetitcolas) [inline doc/Theming.md]: Use a `const` (we are already writing our snippets in ES6)

File: src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            switch (error.status) {
+            case 404:
+               this.handle_404_error();
+            case 403:
+               this.handle_403_error(error);
+            default:
+               this.handle_default_error(error);
+            }
+    	},
+    	handle_404_error: function(event,error){
```
Comment (jpetitcolas) [inline src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js]: I would instead use camelCase.

File: src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+export default function httpErrorService( $state, $translate, notification) {
```
Comment (jpetitcolas) [inline src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js]: CS: no space after opening parenthesis

File: src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+               this.handle_404_error();
+            case 403:
+               this.handle_403_error(error);
+            default:
+               this.handle_default_error(error);
+            }
+    	},
+    	handle_404_error: function(event,error){
+    		 $state.go('ma-404');
+             event.preventDefault();
```
Comment (jpetitcolas) [inline src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js]: What is default behavior?

File: src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    	},
+    	handle_403_error: function(event,error){
+    		 $translate('STATE_FORBIDDEN_ERROR', { message: error.data }).then(text => notification.log(text, { addnCls: 'humane-flatty-error' }));
```
Comment (jpetitcolas) [inline src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js]: I would factorize the `then` part, to get something like `.then(displayError)`.

File: doc/Theming.md (hunk 1 above)
Comment (Kmaschta) [inline doc/Theming.md]: Dunno. But we can make a test on sedy. Can you create an issue?

File: src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js (hunk 5 above)
Comment (Phocea) [inline src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js]: This is linked to the $stateChangeError UI router event, it stops the requested transition to happen

File: doc/Theming.md (hunk 2 above)
Comment (Phocea) [inline doc/Theming.md]: In  that case i can provide the fullblown example with the decorator code inside its own function, but I am not sure that it will help the documentation to be clearer

File: doc/Theming.md (hunk 1 above)
Comment (Phocea) [inline doc/Theming.md]: I am confused here.. what link do I need to change?

File: src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js (hunk 3 above)
Comment (Phocea) [inline src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js]: done

**Comment by Phocea** [comment]
Will add the test for the service tomorrow, was kind of expecting this ! Not fair though, there were none on the original errorHandler to start with ;)

File: doc/Theming.md (hunk 1 above)
Comment (jpetitcolas) [inline doc/Theming.md]: > Http Error Service (../src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js)

to

> [Http Error Service](../src/javascripts/ng-admin/Main/component/provider/HttpErrorService.js)

**Comment by Phocea** [comment]
@jpetitcolas For the unit test, would using $emit directly be enough (this.$rootScope.$emit('$stateChangeError'...) ? This would bypass ui-router completely but I feel that this is the right way to go for this particular service

**Comment by jpetitcolas** [comment]
@Phocea: yes. We just want to catch this specific event. We trust ui-router to emit it correctly on its side. :)

**Comment by Phocea** [comment]
@jpetitcolas @Kmaschta I will close this pull request so I can rebase and redo it with all required files.
Cant do a force push from my remote location at the moment :(
